src/shogun/features/Features.cpp

```diff
@@ -49,7 +49,7 @@ Features::~Features()
 
 void Features::init()
 {
-	set_default_mask(ParameterProperties::READONLY);
+	// set_default_mask(ParameterProperties::READONLY);
```

Comment:
@karlnapf I am thinking if all parameters that are READONLY should be std::optional. This would mean the user can still set a value that is READONLY, if it is not set, and if it is set then it will throw. The current implementation only works when you can create the object via the constructor, so something like this won't work:
```python
hashed_feats=sg.create_features("HashedDocDotFeatures", num_bits=num_bits, 
                                doc_collection=f, tokenizer=tokenizer, 
                                should_normalize=normalize)
```
because it will call the default constructor and then try to put the parameters

Reply:
So this will allow setting values once?
I think this wont be appropriate everywhere, though this is a case where it is.
I guess our "current solution" for this type of probem is offering different factories, but in this case here this doesnt really make sense.
I guess that means we kind of need two types. READONLY for parameters that can never be set from outside (like labels in Machine). And then things like the mask here (not sure what a good name would be). Many of this second type will need callbacks attached to them...

BTW there are also differences for serialization. READONLY would not necessarily be serialized I guess? Or do we actually need a third switch to turn off serialization for particular parameters?

Reply:
>I guess that means we kind of need two types. READONLY for parameters that can never be set from outside (like labels in Machine). And then things like the mask here (not sure what a good name would be). Many of this second type will need callbacks attached to them...

In terms of name it could be something simple `WRITE_ONCE`. We wouldn't need a callback though, we just have the logic implemented like with readonly. And the std::optional would be handled similarly to AutoValue, ie. by the visitor. I assume you mean a callback to check if the value has been set?

>BTW there are also differences for serialization. READONLY would not necessarily be serialized I guess? Or do we actually need a third switch to turn off serialization for particular parameters?

That is a good q. I am not sure how it works rn... I guess it works because we don't serialise the features/labels which are the only parameters with this property?

Reply:
++ for WRITE_ONCE .... re the callback, I think for many cases we will need one (think some statistics are initialized once the value is written), but that is not a problem as we can actually do that

I think for serialization we are currently manually excluding a few parameter properties (like FUNCTION ... not sure about READONLY). Maybe it might be better to have another flag to explicitly disable, as if we rely on say READONLY and that keeps on changing its semantics, we might introduce bugs ...

Reply:
> ++ for WRITE_ONCE .... re the callback, I think for many cases we will need one (think some statistics are initialized once the value is written), but that is not a problem as we can actually do that

Oh I see, I think callbacks are solvable yes, just need to make sure they are called after the value is set, which I think is already the case.

> I think for serialization we are currently manually excluding a few parameter properties (like FUNCTION ... not sure about READONLY). Maybe it might be better to have another flag to explicitly disable, as if we rely on say READONLY and that keeps on changing its semantics, we might introduce bugs ...

OK, I will have a look into this. I will fix this in another PR, and then when that is merged can fix this here.

Reply:
serialization only checks whether the value is visitable... if it is, it'll be serialized

Reply:
ah yes thx, and functions are e.g. not visitable.
The thing is that we sometimes register things that *are* visit-able, but which shouldnt be serialized. This includes for example tables that are created in the ctor and might be exposed via READONLY.